### PR TITLE
Secret Sharing C++ API and first integration in CCF

### PR DESCRIPTION
**Pull request by jumaffre**
This PR relates to 3. and 4. in #51 

- Added a C++ API for `sss` (see `secretshare.h` and unit test).
- Added new `ccf.shares` table, to keep track of the `k_z`-encrypted ledger secrets and the associated member-encrypted shares.
- Started integration with CCF. When the first node starts up, it now encrypts the fresh ledger secret with a new fresh key (`k_z`) that _should_ (*) be split and encrypted for each member. However, since we do not have public-key encryption just yet, we split a dummy variable and record these shares (otherwise anyone with access to the ledger would be able to combine the shares and decrypt the ledger secret).

Next:
- Evercrypt C++ API to do public-key encryption on Ed25519.
- Actually split `k_z` and encrypt the shares with each member's public encryption key.

### CODE REVIEW

File: src/node/nodestate.h [hunk 1]
```diff
@@ -1162,6 +1096,121 @@ namespace ccf
       secrets_view->put(version, secret_set);
     }
 
+    KeyShareInfo generate_key_share_info(
+      const std::vector<ccf::MemberPubInfo>& members_info)
+    {
+      auto share_wrapping_key_raw =
```
Comment (jumaffre) [inline src/node/nodestate.h]: Once we've got https://github.com/microsoft/CCF/issues/793, we should clear this variable once it has been split.

File: src/node/nodestate.h [hunk 2]
```diff
@@ -360,7 +294,7 @@ namespace ccf
           // network
           open_member_frontend();
 
-          if (!create_and_send_request(args, quote))
+          if (!create_and_send_request(args, quote, key_share_info))
```
Comment (jumaffre) [inline src/node/nodestate.h]: I'd forgotten that the first entry was executed by the member frontend. So we generate the encrypted ledger secrets and shares here and pass it to the frontend for execution. This is safe for PBFT as the request recorded in the ledger only contains public info.

File: src/node/nodestate.h [hunk 3]
```diff
@@ -1162,6 +1096,121 @@ namespace ccf
...
+      std::vector<std::vector<uint8_t>> encrypted_shares;
+      for (auto const& s : shares)
+      {
+        encrypted_shares.emplace_back(s.begin(), s.end());
+      }
+
+      return {encrypted_ls.serialise(), encrypted_shares};
+    }
+
+    std::vector<uint8_t> serialize_create_request(
```
Comment (jumaffre) [inline src/node/nodestate.h]: Just moved the following three functions down here to be in the private access specifier.

**Comment by cimetrics** [comment]
shares_table_ccf@[4955 aka 20200214.15](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4955&view=results) vs master ewma over 30 builds from [4560](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4560&view=results) to [4947](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4947&view=results)
images

File: src/node/nodestate.h [hunk 4]
```diff
@@ -1162,6 +1096,121 @@ namespace ccf
...
+        encrypted_ls.hdr.get_iv(), // iv is always 0 here as the share wrapping
+                                   // key is never re-used for encryption
+        network.ledger_secrets->get_secret(1)->master,
+        nullb,
+        encrypted_ls.cipher.data(),
+        encrypted_ls.hdr.tag);
+
+      auto active_members = members_info.size();
+
+      // For now, the secret sharing threashold is set to the number of initial
```
Comment (achamayou) [inline src/node/nodestate.h]: threshold

File: src/node/nodestate.h [hunk 5]
```diff
@@ -1162,6 +1096,121 @@ namespace ccf
...
+      // members
+      size_t k = active_members;
```
Comment (achamayou) [inline src/node/nodestate.h]: number_of_shares is perhaps more explicit than k?

File: src/node/nodestate.h [hunk 6]
```diff
@@ -1162,6 +1096,121 @@ namespace ccf
...
+      }
+
+      create_rpc.params.gov_script = args.config.genesis.gov_script;
+      create_rpc.params.node_cert = node_cert;
+      create_rpc.params.network_cert = network.identity->cert;
+      create_rpc.params.quote = quote;
+      create_rpc.params.public_encryption_key =
+        node_encrypt_kp->public_key_pem().raw();
+      create_rpc.params.code_digest =
+        std::vector<uint8_t>(std::begin(node_code_id), std::end(node_code_id));
```
Comment (achamayou) [inline src/node/nodestate.h]: .begin() and .end()?

File: src/node/secretshare.h [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  // The size of the secret to share is fixed (SECRET_TO_SHARE_LENGTH, 64
+  // bytes). It is up to the caller to either shrink the secret if it is too
+  // long. If the secret to split is shorter than SECRET_TO_SHARE_LENGTH bytes,
+  // the caller should ignore the extra bytes.
+  class SecretSharing
+  {
+  public:
+    static constexpr size_t SECRET_TO_SHARE_LENGTH = sss_MLEN;
+    static constexpr size_t SHARE_LENGTH = sss_SHARE_LEN;
+    static constexpr size_t MAX_NUMBER_SHARES = 255; // As per sss documentation
```
Comment (achamayou) [inline src/node/secretshare.h]: This is now effectively an upper bound on the number of CCF members, we ought to document that, and probably think about solving it.

File: src/node/secretshare.h [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+      }
+
+      if (k == 0 || k > n)
+      {
+        throw std::logic_error(fmt::format("k not in 1-n range (n: {})", n));
+      }
+
+      std::vector<Share> shares(n);
+
+      sss_create_shares(
```
Comment (achamayou) [inline src/node/secretshare.h]: best to avoid C casts as much as possible.

File: src/node/secretshare.h (hunk 7 above)
Comment (achamayou) [reply src/node/secretshare.h]: #851

File: src/node/nodestate.h (hunk 5 above)
Comment (jumaffre) [reply src/node/nodestate.h]: `number_of_shares` is more appropriate for `n` than for `k`. Let's use `threshold` for now. This code is going to be revisited very soon anyway.

**Pull request merged**